crates/solver/src/settlement_access_list.rs

```diff
@@ -14,15 +19,73 @@ pub trait AccessListEstimating: Send + Sync {
         tx: &TransactionBuilder<DynTransport>,
     ) -> Result<AccessList> {
         self.estimate_access_lists(std::slice::from_ref(tx))
-            .await
+            .await?
             .into_iter()
             .next()
             .unwrap()
     }
+    /// The function guarantee the same size and order of input and output containers
     async fn estimate_access_lists(
         &self,
         txs: &[TransactionBuilder<DynTransport>],
-    ) -> Vec<Result<AccessList>>;
+    ) -> Result<Vec<Result<AccessList>>>;
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct NodeAccessList {
+    access_list: AccessList,
+}
+
+#[derive(Debug)]
+pub struct NodeApi {
+    web3: Web3,
+}
+
+impl NodeApi {
+    pub fn new(web3: Web3) -> Self {
+        Self { web3 }
+    }
+}
+
+#[async_trait::async_trait]
+impl AccessListEstimating for NodeApi {
+    async fn estimate_access_lists(
+        &self,
+        txs: &[TransactionBuilder<DynTransport>],
+    ) -> Result<Vec<Result<AccessList>>> {
+        let batch_request = txs
+            .iter()
+            .map(|tx| {
+                let (from, to, data) = resolve_call_request(tx).unwrap_or_default();
```

Comment:
This seems somewhat problematic depending on how often `resolve_call_request()` returns an error.
First we send a request which will fail (right?) and the final `Vec<Result<_>>` hides the fact that we weren't able to resolve this request in the first place and instead reports this as a web3 or parsing error.

Ideally we would only send the requests which we expect to return sensible results and propagate the error of `resolve_call_request()`. This is how I handled the pattern of `(1) filter initial errors (2) send only sensible requests (3) merge results` in the price estimator code. This is the most elegant I could think of but maybe there is still room for improvement:
```rust
async fn estimate_access_lists(
    &self,
    txs: &[TransactionBuilder<DynTransport>],
) -> Result<Vec<Result<AccessList>>> {
    let batch_request = txs
        .iter()
        .map(|tx| -> Result<_, anyhow::Error> {
            let (from, to, data) = resolve_call_request(tx)?;
            let request = CallRequest {
                from: Some(from),
                to: Some(to),
                data: Some(data),
                ..Default::default()
            };
            let params = helpers::serialize(&request);
            let (id, request) = self
                .web3
                .transport()
                .prepare("eth_createAccessList", vec![params]);
            Ok((id, request))
        })
        .collect::<Vec<_>>();

    // send_batch guarantees the size and order of the responses to match the requests
    let mut batch_response = self
        .web3
        .transport()
        .send_batch(batch_request.iter().flatten().cloned())
        .await?
        .into_iter();

    Ok(batch_request
        .into_iter()
        // merge results of unresolved call requests with responses of resolved requests
        .map(|req| match req {
            // error during `resolve_call_request()`
            Err(e) => Err(e),
            Ok(_req) => match batch_response.next().unwrap() {
                Ok(response) => serde_json::from_value::<NodeAccessList>(response)
                    // error parsing the response
                    .context("unexpected response format")
                    .map(|response| response.access_list),
                // error during transport
                Err(err) => Err(anyhow!("web3 error: {}", err)),
            },
        })
        .collect())
}
```